logger/configurer/configurer.go

```diff
@@ -0,0 +1,137 @@
+package configurer
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/coreos/go-etcd/etcd"
+	"github.com/deis/deis/logger/drain"
+	"github.com/deis/deis/logger/storage"
+	"github.com/deis/deis/logger/syslogish"
+)
+
+// DefaultDrainURI is exported so its value can be specified by an external agent.  In practical
+// use, main.go sets this after parsing command line flags in order to support a legacy behavior
+// that that allows default drain uri to be specified using a drain-uri flag.
+var DefaultDrainURI string
+
+// Configurer takes responsibility for dynamically reconfiguring a syslogish.Server based on
+// changes in etcd.
+type Configurer struct {
+	etcdClient                *etcd.Client
+	etcdPath                  string
+	ticker                    *time.Ticker
+	syslogishServer           *syslogish.Server
+	running                   bool
+	currentStorageAdapterType string
+	currentDrainURL           string
+}
+
+// NewConfigurer returns a pointer to a new Configurer instance.
+func NewConfigurer(etcdHost string, etcdPort int, etcdPath string, configInterval int,
+	syslogishServer *syslogish.Server) (*Configurer, error) {
+	etcdClient := etcd.NewClient([]string{fmt.Sprintf("http://%s:%d", etcdHost, etcdPort)})
+	ticker := time.NewTicker(time.Duration(configInterval) * time.Second)
+	configurer := &Configurer{
+		etcdClient:      etcdClient,
+		etcdPath:        etcdPath,
+		syslogishServer: syslogishServer,
+		ticker:          ticker,
+	}
+
+	// Support legacy behavior that allows default drain uri to be specified using a drain-uri flag
+	if _, err := etcdClient.Get(etcdPath+"/drain", false, false); err != nil {
+		etcdErr, ok := err.(*etcd.EtcdError)
+		// Error code 100 is key not found
+		if ok && etcdErr.ErrorCode == 100 {
+			configurer.setEtcd("/drain", DefaultDrainURI)
+		} else {
+			log.Println(err)
+		}
+	}
+
+	return configurer, nil
+}
+
+// Start begins the configurer's main loop.
+func (c *Configurer) Start() {
+	// Should only ever be called once
+	if !c.running {
+		c.running = true
+		go c.configure()
+		log.Println("configurer running")
+	}
+}
+
+func (c *Configurer) configure() {
+	for {
+		<-c.ticker.C
+		c.manageStorageAdapter()
+		c.manageDrain()
+	}
+}
+
+func (c *Configurer) manageStorageAdapter() {
+	newStorageAdapterType, err := c.getEtcd("/storageAdapterType", "file")
+	if err != nil {
+		log.Println("configurer: Error retrieving storage adapter type from etcd.  Skipping.", err)
+		return
+	}
+	if newStorageAdapterType != c.currentStorageAdapterType {
```

Comment:
-- nitpick mode on --

I'm not a big fan for a long if statement, this would look much nicer the other way around:

```
ffunc (c *Configurer) manageStorageAdapter() {
	newStorageAdapterType, err := c.getEtcd("/storageAdapterType", "file")
	if err != nil {
		log.Println("configurer: Error retrieving storage adapter type from etcd.  Skipping.", err)
		return
	}
	if newStorageAdapterType == c.currentStorageAdapterType {
		return
	}
	newStorageAdapter, err := storage.NewAdapter(newStorageAdapterType)
	if err != nil {
		log.Println("configurer: Error creating new storage adapter.  Skipping.", err)
		return
	}
	c.syslogishServer.SetStorageAdapter(newStorageAdapter)
	c.currentStorageAdapterType = newStorageAdapterType
	log.Printf("configurer: Activated new storage adapter: %s", newStorageAdapterType)
}
```

-- nitpick mode off --